rstudio/rstudio
# Change DB File Scheme

### Intent

This addresses concerns that were raised on how to manage Database compatibility between versions. In particular this chart should hold true (with regards to what is supported):

                     | Dailies | Previews | Releases |
    -----------------+---------+----------+----------+
     Upgrade From    | Yes     | Yes      | Yes      |
     Upgrade to      | Yes     | Yes      | Yes      |
     Downgrade to    | No      | No       | Yes      |
     Downgrade from  | Yes     | Yes      | Yes      |

### Approach

* There is a single `CreateTables` file (one for postgres, one for sqlite) for each release. 
  * This file is used for first installation. It may be changed every release.
* There is an `AlterTables` file (one for postgres, one for sqlite) for each release.
  * Once a version is released, this file may not be modified.
  * Actions in this file should be idempotent to facilitate the `daily 1 > daily x > preview > release` upgrade path.
  * Changes to this file require an update the `DATE` portion of the file.

### Automated Tests

All server_core unit tests pass.

### QA Notes

* It's important to check both new installations and upgrades here.
* In the future, there will be other changes to the `AlterTables` file. Once that happens, it will be important to verify that this daily updates smoothly to the new daily.
* Both Postgres & sqlite dbs will need to be tested.
* We should also test that downgrading to JR or WB doesn't break anything.

### Checklist
